### PR TITLE
Close #13137 the hard way; reworked UniqueIDs in virus2

### LINKED ISSUE #13137: 2 viruses with the same name but 2 different verosity + disease analyzer
(WEB REPORT BY: pvak REMOTE: 172.93.109.202:7777)
>Revision (Should be above if you're viewing this from ingame!)

>General description of the issue
if 2 viruses have the same name but different small stats when you put them in the disease analyzer and print the paper it ll print 2 papers of the 2 viruses

>What you expected to happen
1 paper

>What actually happened
2 papers

>Steps to reproduce if possible

have 2 viruses with slight different stats but same synthoms and name 
but in the analyzer
print
2 papers istead of 1

### PR DESCRIPTION
**Pull request by despotate**
Closes #13137 the ~~lazy~~ hard way

### CODE REVIEW

**Comment by ComicIronic** [comment]
Given that this doesn't make a new ID, it's not going to solve the bug properly. Use a global var to store the last virus ID and make a proc to generate an incrementing ID.

**Comment by despotate** [comment]
How does it not solve the bug? It makes minormutate generate a new ID the same as majormutate so the viruses have different names. Thus the analyser won't spit out multiple dishes/papers as they don't have the same name.

I'd also like you to elaborate on what you mean with "last virus ID" and generating an incrementing ID. Virus IDs are just random numbers from 1-10000, there aren't any other rules constraining them. If I just incremented the ID, then we'd still have the same problem as two copies of one dish would get different mutations but receive the same new ID, and thus get popped/papered together in the analyser.

**Comment by Exxion** [comment]
He means have a global var that increments every time a virus is created or mutated, and assign every virus's ID to it on creation/mutation

**Comment by ComicIronic** [comment]
I can still have a virus mutate and have the same ID before, you've just made the bug much rarer.

**Comment by despotate** [comment]
Which is why I said I'm solving this bug "the lazy way", ie, the way majormutate currently does it.

Very unlikely to have a virus conflict though.

Regardless, I'm gonna add the global var - but where do I add it? _DEFINES/global.dm?

**Comment by Exxion** [comment]
Nah
Just add it to the virus definition
`var/global/idholder` or something

**Comment by PJB3005** [comment]
Problem with the "proper" way of incrementing a global is that it's sort of immersion breaking. You'd be able to tell at a glance how much different viruses there have been because of the ID of a new mutation you just did.

Best solution would be to make a list and add any IDs to that list, then don't allow dupes by checking if it's in that list...

Gonna be slow though.

**Comment by Exxion** [comment]
You could start with a random ID and increment the value by a random number each time

**Comment by ComicIronic** [comment]
Yeah, if you pick a random number between 0 and 100000 exclusive, and every time you increment the ID by that number and do `% 100000`, you'd get a random-looking ID always in the range until you ran out of IDs.

**Comment by Exxion** [comment]
Only if said random number is mutually prime with 100000

**Comment by despotate** [comment]
We just need to find a one-to-one mapping of digits from 1 to 10000 to "random" numbers from 1 to 10000 that changes each game.

**Comment by Exxion** [comment]
hashing lol

**Comment by despotate** [comment]
Never done it. What should I do to implement it in virus2?

**Comment by PJB3005** [comment]
I was about to mention hashing but that'd probably make hash collision a hell and problem.

**Comment by despotate** [comment]
Can we just leave it at generating a random number every time? Maybe if we raised the number of IDs from 10000 to 1000000?

**Comment by Exxion** [comment]
I'd still go with my first suggestion
Generate a random starting ID between like 1 and 5000 and every time you generate a new ID, add a random number between, for example, 1 and 50 to it
So people could tell what ORDER viruses were made in, but couldn't necessarily tell how many were made between two given IDs

**Comment by despotate** [comment]
Yeah with a high enough upper limit to this random addition you wouldn't be able to powergame it

I'll go with that

**Comment by Probe1** [comment]
Isn't coding fun?

This should be an easy fix..FUCK

**Comment by ghost** [comment]
@ComicIronic The random numbers on the end of monkey names, zombie names, etc. is just that, random.

It doesn't go off a globally incrementing variable. Leave it as is, it's a 1 in 10000

**Comment by despotate** [comment]
The real "easy fix" would be to up the number to 1000000 (to please the MUH COLLISIONS crowd) and let minormutate change it

I don't know why we're doing this complicated shit that gives powergamers a bit more information than they need (order of randomly generated viruses that were spawned outside the lab)

**Comment by PJB3005** [comment]
Just put the number range large enough, make a global list of used IDs, then try to generate numbers until you found nothing conflicting.

**Comment by despotate** [comment]
How about we just don't give a shit about solving literally one-in-a-million problems by adding more work and creating larger problems?

This is starting to remind me of the Voter ID argument in the United States.

**Comment by ComicIronic** [comment]
There are rarely any bad effects from zombie or monkey name collision. The issue I have with this is that enough viruses are generated that based on the birthday problem this is going to become a bug that still happens often enough to get a report, and rarely enough to be difficult to reproduce. If the issue is that viruses need unique IDs, then let's fix that issue, not half-arse it because you people are lazy.

**Comment by ghost** [comment]
>You people

He just wants to fix a bug, not rewrite half the damn codebase

Hell, have it so that you can set the unique ID if needs be via the computer, so should somebody be infected with it while it's logged in the database, they'll see 'Oh, it's beneficial virus/Honkstains/Gayniggerstonguemyanus syndrome' rather than a string of numbers between 0 and 10000

**Comment by despotate** [comment]
Look if we add two more digits to the unique ID and have 100 different viruses generated per game, then there's a 0.5% chance of a collision per game https://www.wolframalpha.com/input/?i=1-e%5E((-100%5E2)%2F(2x1000000)) so I really can't be bothered to solve this the "right" way other than to clean up the code some more

Skullyton, you CAN actually rename the virus through the laptop. No one ever does it though

**Comment by Probe1** [comment]
I mean no one really plays Viro.  But everyone I've taught knows to name their beneficial viruses to lower the chance they're lynched.

**Comment by despotate** [comment]
My final demands are:
- Let me bandaid it, increase the amount of digits by two, and allow minormutate to change the ID, OR
- Tell me how to implement a proper, non-colliding hashing function and I'll do that

**Comment by PJB3005** [comment]
Just do the former and keep a global list of used IDs, if it collides try again (up to 5 times to prevent looping issues) with a fresh ID.

**Comment by despotate** [comment]
Will do

File: code/modules/virus2/disease2.dm [hunk 1]
```diff
@@ -16,11 +17,17 @@ var/global/list/disease2_list = list()
...
 /datum/disease2/disease/New(var/notes="No notes.")
-	uniqueID = rand(0,10000)
+	generate_uniqueID()
 	log += "<br />[timestamp()] CREATED - [notes]<br>"
-	disease2_list["[uniqueID]"] = src
 	..()
 
+/datum/disease2/disease/proc/generate_uniqueID()
+	for (var/x = 1 to 5)
+    	uniqueID = rand(...)
```
Comment (ComicIronic) [inline code/modules/virus2/disease2.dm]: This has to be indented under the for.

Comment (despotate) [reply code/modules/virus2/disease2.dm]: I copy pasted from discord and apparently it added a mix of spaces + tabs
and the spaces get ignored

Comment (PJB3005) [reply code/modules/virus2/disease2.dm]: ...Did you really copypaste the code this literally?

How the shit does `rand(...)` even compile?

When I wrote that snippet for you I assumed you'd fill in the rand...

File: code/modules/virus2/disease2.dm [hunk 2]
```diff
@@ -1,4 +1,5 @@
 var/global/list/disease2_list = list()
+var/global/list/virusDB = list()
```
Comment (PJB3005) [inline code/modules/virus2/disease2.dm:2]: Remove `/global` here.

Comment (despotate) [reply code/modules/virus2/disease2.dm:2]: That line was there before, I simply moved it from the middle of the file to the top.

I think it's used in the medical laptop and HUD.

Apparently oldcoders made TWO different virus databases and refused to consider using either one to generate new IDs.

File: code/modules/virus2/disease2.dm (hunk 1 above)
Comment (despotate) [reply code/modules/virus2/disease2.dm]: easy to skim over code that looks correct
I did realize it almost immediately after of course https://github.com/d3athrow/vgstation13/pull/13141/commits/5beeb798a86f74419f2df66d8a92599f8a9627ab

File: code/modules/virus2/disease2.dm (hunk 2 above)
Comment (PJB3005) [reply code/modules/virus2/disease2.dm:2]: Eh OK then.

Hut general rule of thumb: don't keep `/global` for global variables. It doesn't do anything.

**Comment by despotate** [comment]
FEEL THE FUCK FREE TO REVERT THIS AAAAAAAAAAAAAAAAAAAAAAH I WARNED YOU ABOUT DOING IT THE COMPLICATED WAY

**Pull request merged**